src/main/java/frc/trigon/robot/poseestimation/poseestimator/PoseEstimator.java

```diff
@@ -113,41 +134,179 @@ private void logTargetPath() {
         PathPlannerLogging.setLogTargetPoseCallback((pose) -> Logger.recordOutput("TargetPPPose", pose));
     }
 
+    /**
+     * Sets the odometry position at a given timestamp.
+     *
+     * @param swerveModulePositions the positions of each swerve module
+     * @param gyroHeading           the heading of the gyro
+     * @param timestamp             the timestamp of the odometry observation
+     */
+    private void addOdometryObservation(SwerveModulePosition[] swerveModulePositions, Rotation2d gyroHeading, double timestamp) {
+        updateOdometryPositions(swerveModulePositions, gyroHeading);
+
+        final Twist2d newOdometryPoseDifference = calculateNewOdometryPoseDifference(swerveModulePositions, gyroHeading);
+        updateRobotPosesFromNewOdometryPoseDifference(newOdometryPoseDifference, timestamp);
+    }
+
     private void updateFromVision() {
-        getViableVisionObservations().stream()
-                .sorted(Comparator.comparingDouble(PoseEstimator6328.VisionObservation::timestamp))
-                .forEach(poseEstimator6328::addVisionObservation);
+        updateAllCameras();
+        sortCamerasByLastTargetTimestamp();
+        addVisionObservations();
     }
 
-    private List<PoseEstimator6328.VisionObservation> getViableVisionObservations() {
-        List<PoseEstimator6328.VisionObservation> viableVisionObservations = new ArrayList<>();
+    private void updateAllCameras() {
         for (AprilTagCamera aprilTagCamera : aprilTagCameras) {
-            final PoseEstimator6328.VisionObservation visionObservation = getVisionObservation(aprilTagCamera);
-            if (visionObservation != null)
-                viableVisionObservations.add(visionObservation);
+            aprilTagCamera.update();
         }
-        return viableVisionObservations;
     }
 
-    private PoseEstimator6328.VisionObservation getVisionObservation(AprilTagCamera aprilTagCamera) {
-        aprilTagCamera.update();
-        if (!aprilTagCamera.hasNewResult())
-            return null;
+    private void sortCamerasByLastTargetTimestamp() {
+        VisionObservationSorter.quickSortByDoubleValue(aprilTagCameras, (aprilTagCamera) -> ((AprilTagCamera) aprilTagCamera).getLatestResultTimestampSeconds());
+    }
+
+    private void updateFromT265() {
+        for (AprilTagCamera aprilTagCamera : aprilTagCameras)
+            if (aprilTagCamera.isWithinBestTagRangeForSolvePNP())
```

Comment:
I might be wrong, but I thought we were only using solvePNP to get the initial pose at the start of the match. Not everytime we are close enough to a tag. The whole point of the t265 is that it's instead of it.